### PR TITLE
add clusterrole bind/view for global set

### PR DESCRIPTION
**Pull request by ldpliu**
Signed-off-by: ldpliu <daliu@redhat.com>
https://github.com/stolostron/backlog/issues/22995

### CODE REVIEW

File: pkg/controllers/clusterset/clustersetmapper/clusterset_controller.go [hunk 1]
```diff
@@ -221,15 +217,23 @@ func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Resu
 
 	//add clusterset clusterrole
 	adminroleName := utils.GenerateClustersetClusterroleName(clusterset.Name, "admin")
-	adminRole := buildAdminRole(clusterset.Name, adminroleName)
+	adminRole := clustersetutils.BuildAdminRole(clusterset.Name, adminroleName)
```
Comment (qiujian16) [inline pkg/controllers/clusterset/clustersetmapper/clusterset_controller.go:220]: I think you should put line 219 in the buildxxx func.

File: pkg/controllers/clusterset/clustersetmapper/clusterset_controller.go [hunk 2]
```diff
@@ -221,15 +217,23 @@ func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Resu
...
 	err = utils.ApplyClusterRole(r.kubeClient, adminRole)
 	if err != nil {
 		klog.Warningf("will reconcile since failed to create/update clusterrole %v, %v", clusterset.Name, err)
 		return ctrl.Result{}, err
 	}
 
+	bindroleName := utils.GenerateClustersetClusterroleName(clusterset.Name, "bind")
+	bindRole := clustersetutils.BuildBindRole(clusterset.Name, bindroleName)
+	err = utils.ApplyClusterRole(r.kubeClient, bindRole)
```
Comment (qiujian16) [inline pkg/controllers/clusterset/clustersetmapper/clusterset_controller.go:229]: I think we should apply all clusterroles, and return the aggregate err

File: pkg/controllers/clusterset/globalclusterset/global_clusterset_controller.go [hunk 3]
```diff
@@ -0,0 +1,167 @@
+package globalclusterset
+
+import (
+	"context"
```
Comment (qiujian16) [inline pkg/controllers/clusterset/globalclusterset/global_clusterset_controller.go:4]: why we cannot do it in clusterset_controller?

File: pkg/utils/clusterset/clusterset.go [hunk 4]
```diff
@@ -76,3 +82,112 @@ func GenerateClustersetSubjects(cache *cache.AuthCache) map[string][]rbacv1.Subj
 
 	return clustersetToSubjects
 }
+
+// BuildAdminRoleRules builds the clustesetadminroles
+func BuildAdminRole(clustersetName, clusteroleName string) *rbacv1.ClusterRole {
```
Comment (qiujian16) [inline pkg/utils/clusterset/clusterset.go:87]: more comments will be helpful to describe the permission for each clusterrole

Comment (ldpliu) [reply pkg/utils/clusterset/clusterset.go]: done

File: pkg/controllers/clusterset/globalclusterset/global_clusterset_controller.go (hunk 3 above)
Comment (ldpliu) [reply pkg/controllers/clusterset/globalclusterset/global_clusterset_controller.go]: moved

File: pkg/controllers/clusterset/clustersetmapper/clusterset_controller.go (hunk 2 above)
Comment (ldpliu) [reply pkg/controllers/clusterset/clustersetmapper/clusterset_controller.go]: done

File: pkg/controllers/clusterset/clustersetmapper/clusterset_controller.go (hunk 1 above)
Comment (ldpliu) [reply pkg/controllers/clusterset/clustersetmapper/clusterset_controller.go]: done

File: pkg/controllers/clusterset/clustersetmapper/clusterset_controller.go [hunk 5]
```diff
@@ -243,24 +227,66 @@ func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Resu
 	return ctrl.Result{}, nil
 }
 
+//applyClusterSetClusterRoles apply the clusterset clusterrole(admin/bind/view)
+func (r *Reconciler) applyClusterSetClusterRoles(clustersetName string) error {
+	errs := []error{}
+	if clustersetName != clustersetutils.GlobalSetName {
```
Comment (qiujian16) [inline pkg/controllers/clusterset/clustersetmapper/clusterset_controller.go:233]: can you just use selectorType here?

File: pkg/utils/clusterset/clusterset.go [hunk 6]
```diff
@@ -11,10 +11,15 @@ import (
 )
 
 const (
-	ClusterSetRole string = "cluster.open-cluster-management.io/role"
-
-	DefaultSetName = "default"
-	GlobalSetName  = "global"
+	ClusterSetLabel             string = "cluster.open-cluster-management.io/clusterset"
```
Comment (qiujian16) [inline pkg/utils/clusterset/clusterset.go:14]: you do not need this, right?

File: pkg/controllers/clusterset/clustersetmapper/clusterset_controller.go (hunk 5 above)
Comment (ldpliu) [reply pkg/controllers/clusterset/clustersetmapper/clusterset_controller.go]: done

File: pkg/controllers/clusterset/clustersetmapper/clusterset_controller.go [hunk 7]
```diff
@@ -209,30 +200,23 @@ func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Resu
...
 		return ctrl.Result{}, err
 	}
 
-	viewroleName := utils.GenerateClustersetClusterroleName(clusterset.Name, "view")
-	viewRole := buildViewRole(clusterset.Name, viewroleName)
-	err = utils.ApplyClusterRole(r.kubeClient, viewRole)
-	if err != nil {
-		klog.Warningf("will reconcile since failed to create/update clusterrole %v, %v", clusterset.Name, err)
-		return ctrl.Result{}, err
+	if clusterset.Name == clustersetutils.GlobalSetName {
```
[the remark this answers is not in the document]
Comment (ldpliu) [reply pkg/controllers/clusterset/clustersetmapper/clusterset_controller.go:218]: The following code in the reconcile func generate the clustersetMapper, which used to propagate the clusterset permission to managedcluster/cluster ns/clusterpool/clusterclaims/clusterdeployments. They should not work for gloabl set.

File: pkg/utils/clusterset/clusterset.go (hunk 6 above)
Comment (ldpliu) [reply pkg/utils/clusterset/clusterset.go]: removed.

File: pkg/controllers/clusterset/clustersetmapper/clusterset_controller.go [hunk 8]
```diff
@@ -243,24 +227,66 @@ func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Resu
...
+			errs = append(errs, err)
+		}
+	}
+
+	bindRole := clustersetutils.BuildBindRole(clustersetName)
+	viewRole := clustersetutils.BuildViewRole(clustersetName)
+
+	//Add all clusters view permission to global set bind and view clusterrole
+	if clustersetName == clustersetutils.GlobalSetName {
+		clustersetutils.AddAllClustersViewPermission(bindRole)
```
[the remark this answers is not in the document]
Comment (qiujian16) [reply pkg/controllers/clusterset/clustersetmapper/clusterset_controller.go:247]: hrm, I thought view role clusterset can view all clusters in the set..

File: pkg/controllers/clusterset/clustersetmapper/clusterset_controller.go [hunk 9]
```diff
@@ -243,24 +229,66 @@ func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Resu
...
 		klog.Warningf("will reconcile since failed to delete clusterrole. clusterset: %v, err: %v", clustersetName, err)
 		return err
 	}
+
 	err = utils.DeleteClusterRole(r.kubeClient, utils.GenerateClustersetClusterroleName(clustersetName, "view"))
 	if err != nil {
 		klog.Warningf("will reconcile since failed to delete clusterrole. clusterset: %v, err: %v", clustersetName, err)
 		return err
 	}
+	if clustersetName != clustersetutils.GlobalSetName {
```
Comment (qiujian16) [inline pkg/controllers/clusterset/clustersetmapper/clusterset_controller.go:281]: this is weird doc is needed

File: pkg/controllers/clusterset/clustersetmapper/clusterset_controller.go [hunk 10]
```diff
@@ -243,24 +229,66 @@ func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Resu
...
+			klog.Warningf("will reconcile since failed to create/update admin clusterrole %v, %v", clusterset.Name, err)
+			errs = append(errs, err)
+		}
+	}
+
+	bindRole := clustersetutils.BuildBindRole(clusterset.Name)
+	viewRole := clustersetutils.BuildViewRole(clusterset.Name)
+
+	//Add all clusters view permission to global set bind and view clusterrole
+	if clusterset.Name == clustersetutils.GlobalSetName {
```
Comment (qiujian16) [inline pkg/controllers/clusterset/clustersetmapper/clusterset_controller.go:248]: this is still not clear to me.

Comment (ldpliu) [reply pkg/controllers/clusterset/clustersetmapper/clusterset_controller.go]: done

File: pkg/controllers/clusterset/clustersetmapper/clusterset_controller.go (hunk 9 above)
Comment (ldpliu) [reply pkg/controllers/clusterset/clustersetmapper/clusterset_controller.go]: done

**Review by qiujian16 — COMMENTED** [review]
/approve
/lgtm

**Pull request merged**